src/sing/stateMachine/sequencerStateMachine.ts

```diff
@@ -0,0 +1,269 @@
+      readonly cursorPos: PositionOnSequencer;
+    }
+  | {
+      readonly targetArea: "Note";
+      readonly mouseEvent: MouseEvent;
+      readonly cursorPos: PositionOnSequencer;
+      readonly note: Note;
+    }
+  | {
+      readonly targetArea: "NoteLeftEdge";
+      readonly mouseEvent: MouseEvent;
+      readonly cursorPos: PositionOnSequencer;
+      readonly note: Note;
+    }
+  | {
+      readonly targetArea: "NoteRightEdge";
+      readonly mouseEvent: MouseEvent;
+      readonly cursorPos: PositionOnSequencer;
+      readonly note: Note;
+    };
+
+type ComputedRefs = {
+  readonly snapTicks: ComputedRef<number>;
+  readonly editTarget: ComputedRef<SequencerEditTarget>;
+};
+
+type Refs = {
+  readonly nowPreviewing: Ref<boolean>;
+  readonly previewNotes: Ref<Note[]>;
+  readonly guideLineTicks: Ref<number>;
+};
+
+type StoreActions = {
+  readonly deselectAllNotes: () => void;
+  readonly commandAddNotes: (notes: Note[]) => void;
+  readonly selectNotes: (noteIds: NoteId[]) => void;
+  readonly playPreviewSound: (noteNumber: number, duration?: number) => void;
+};
+
+type Context = ComputedRefs & Refs & { readonly storeActions: StoreActions };
+
+type State = IdleState | AddNoteState;
+
+const getGuideLineTicks = (
+  cursorPos: PositionOnSequencer,
+  context: Context,
+) => {
+  const cursorTicks = cursorPos.ticks;
+  const snapTicks = context.snapTicks.value;
+  // NOTE: 入力を補助する線の判定の境目はスナップ幅の3/4の位置
+  return Math.round(cursorTicks / snapTicks - 0.25) * snapTicks;
+};
+
+class IdleState implements IState<State, Input, Context> {
+  readonly id = "idle";
+
+  onEnter() {}
+
+  process({
+    input,
+    context,
+    setNextState,
+  }: {
+    input: Input;
+    context: Context;
+    setNextState: (nextState: State) => void;
+  }) {
+    const mouseButton = getButton(input.mouseEvent);
+    if (input.targetArea === "SequencerBody") {
+      context.guideLineTicks.value = getGuideLineTicks(
+        input.cursorPos,
+        context,
+      );
+      if (input.mouseEvent.type === "mousedown") {
+        // TODO: メニューが表示されている場合はメニュー非表示のみ行いたい
+        if (context.editTarget.value === "NOTE") {
+          if (!isSelfEventTarget(input.mouseEvent)) {
+            return;
+          }
+          if (mouseButton === "LEFT_BUTTON") {
+            if (
+              input.cursorPos.ticks < 0 ||
+              input.cursorPos.noteNumber < 0 ||
+              input.cursorPos.noteNumber > 127
+            ) {
+              return;
+            }
+            setNextState(new AddNoteState(input.cursorPos));
+          }
+        }
+      }
+    }
+  }
+
+  onExit() {}
+}
+
+class AddNoteState implements IState<State, Input, Context> {
+  readonly id = "addNote";
+
+  private readonly cursorPosAtStart: PositionOnSequencer;
+
+  private currentCursorPos: PositionOnSequencer;
+  private innerContext:
+    | {
+        noteToAdd: Note;
+        previewRequestId: number;
+        executePreviewProcess: boolean;
```

Comment:
should付けると更に意味がわかりやすいかも？
```suggestion
        shouldExecutePreviewProcess: boolean;
```